Wrappers/Python/cil/optimisation/operators/WaveletOperator.py

```diff
@@ -0,0 +1,185 @@
+# -*- coding: utf-8 -*-
+#  Copyright 2019 United Kingdom Research and Innovation
+#  Copyright 2019 The University of Manchester
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+#
+# Authors:
+# CIL Developers, listed at: https://github.com/TomographicImaging/CIL/blob/master/NOTICE.txt
+
+import numpy as np
+import pywt # PyWavelets module
+
+from cil.optimisation.operators import LinearOperator
+from cil.framework import VectorGeometry
+
+
+###############################################################################
+###############################################################################
+########################## Discrete Wavelet Transform #########################
+###############################################################################
+###############################################################################
+
+class WaveletOperator(LinearOperator):
+    
+    r'''                  
+        Computes forward or inverse (adjoint) discrete wavelet transform of the input
+        
+        :param domain_geometry: Domain geometry for the WaveletOperator
+
+        [OPTIONAL PARAMETERS]
+        :param range_geometry: Output geometry for the WaveletOperator. 
+            Default = domain_geometry with the right coefficient array size deduced from pywt
+        :param level: integer for decomposition level.
+            Default = log_2(min(shape(axes))), i.e. the maximum number of accurate downsamplings possible
+        :type wname: string label for wavelet used.
+            Default = "haar"
+        :type axes: range of ints to define the dimensions to decompose along. Note that channel is the first dimension:
+            For example, spatial DWT is given by axes=range(1,3) and channelwise DWT is axes=range(1)
+            Default = None, meaning all dimensions are transformed. Same as axes=range(ndim)
+        :param moments: integer for number of vanishing moments.
+            Default = Known for Daubechies, None for others
+        
+     '''     
+       
+    def __init__(self, domain_geometry, 
+                       range_geometry=None, 
+                       level = None,
+                       wname = "haar",
+                       axes = None):
```

Comment:
This should use the `dimension_label` or `correlation` as in `GradientOperator`

Reply:
New `correlation` kwarg has been introduced